### DISCUSSION TITLE
API rule to only allow a user to add or remove themself from a relation.

### DISCUSSION THREAD

**Discussion by EricApostal** [pocketbase/pocketbase#3806] [Q&A] [answered]
I am working on an event system where users can sign up for events, and also remove their sign up from such event. This is done with:

```dart
// sign up
await.collection('events').update(id, body: {'attendees+': id});

// remove sign up
await.collection('events').update(id, body: {'attendees-': id});
```
As it stands currently, it's possible for users to spoof the "attendees" field with another's ID. I attempted to add an API rule that required that the requester's ID be in the updated field, but this breaks if they send multiple values, or if they are attempting to remove their sign up.

How could I implement a rule to prevent a user from signing up / removing the signup of someone who isn't authenticated as the user they are modifying?

Thanks

**Comment by ganigeorgiev** [comment, marked as answer]
Currently I don't think it is possible without extra joins or custom hook to do what you want with the existing collections structure.

An alternative approach depending on your use case could be to have a more traditional SQL structure and move the attendees relation in a separate collection, for example `eventAttendee` that will have 2 single `relation` fields - `event` and `user` pointing to their respective collections. You can also define a unique constraint consisting of the 2 fields to prevent a single user to be added twice for the same event.
Note that [back/indirect relations](https://pocketbase.io/docs/working-with-relations/#back-relation-expand) expand is supported and for example if you want to fetch an event with all attendees expanded this could look like:

```js
// adding a new attendee
await.collection('eventAttendee').create({
    event: "EVENT_ID",
    user: "USER_ID",
})

// removing attendee
const rel = await.collection('eventAttendee').getFirstListItem(pb.filter("user={:id}", { id: "USER_ID"}));
await.collection('eventAttendee').delete(rel.id);

// get an event with all attendees
await.collection('events').getOne('EVENT_ID', {
    expand: "eventAttendee(event).user"
});
```
In the api rules back relations are not supported yet, but you can use `@collection.*` to perform joins if you need to perform extra checks on some back relation field.

---

With that said, if you don't want to change your existing collection structure, then you can define a custom hook with Go or JS where you can perform the validation. Here is an example with the [JS pb_hooks](https://pocketbase.io/docs/js-overview/):

```js
// pb_hooks/main.pb.js

// note: I'm not sure if users are allowed to create events, if not remove it
onRecordBeforeCreateRequest((e) => {
    // skip for admin
    if (e.httpContext.get("admin")) {
        return
    }

    // retrieve the current authenticated user
    const authRecord = e.httpContext.get("authRecord")

    // retrieve the submitted attendees
    const attendees = e.record.get("attendees");

    // an attendee can add only themself
    if (attendees.length && (attendees.length > 1 || attendees[0] != authRecord.id)) {
        throw new BadRequestError("Invalid event data", {
            "attendees": new ValidationError("invalid_attendees_value", "An attendee is not allowed to add other attendees."),
        })
    }
}, "events")

onRecordBeforeUpdateRequest((e) => {
    // skip for admin
    if (e.httpContext.get("admin")) {
        return
    }

    // retrieve the current authenticated user
    const authRecord = e.httpContext.get("authRecord")

    // retrieve the old attendees list
    const oldAttendees = e.record.originalCopy().get("attendees");

    // retrieve the current attendees list
    const newAttendees = e.record.get("attendees");

    // an attendee can add only themself
    const added = newAttendees.filter((a) => !oldAttendees.includes(a));
    if (added.length && (added.length > 1 || added[0] != authRecord.id)) {
        throw new BadRequestError("Invalid event data", {
            "attendees": new ValidationError("invalid_attendees_value", "An attendee is not allowed to add other attendees."),
        })
    }

    // an attendee can remove only themself
    const removed = oldAttendees.filter((a) => !newAttendees.includes(a));
    if (removed.length && (removed.length > 1 || removed[0] != authRecord.id)) {
        throw new BadRequestError("Invalid event data", {
            "attendees": new ValidationError("invalid_attendees_value", "An attendee is not allowed to remove other attendees."),
        })
    }
}, "events")
```

**Reply by EricApostal** [reply]
Thanks for the awesome response, this helps quite a lot! Going with a hook seems like my best bet here.
